### PR TITLE
Make def names and HIR names consistent.

### PR DESCRIPTION
**Pull request by cjgillot**
The name in the `DefKey` is interned to create the `DefId`, so it does not
require any query to access.  This can be leveraged to avoid a few useless
HIR accesses for names.

~In order to achieve that, generic parameters created from universal
impl-trait are given the pretty-printed ast as a name, instead of
`{{opaque}}`.~

~Drive-by: the `TyCtxt::opt_item_name` used a dummy span for non-local
definitions.  We have access to `def_ident_span`, so we use it.~

### CODE REVIEW

File: src/test/ui/issues/issue-21763.stderr [hunk 1]
```diff
@@ -7,8 +7,16 @@ LL |     foo::<HashMap<Rc<()>, Rc<()>>>();
    = help: within `(Rc<()>, Rc<()>)`, the trait `Send` is not implemented for `Rc<()>`
    = note: required because it appears within the type `(Rc<()>, Rc<()>)`
    = note: required because of the requirements on the impl of `Send` for `hashbrown::raw::RawTable<(Rc<()>, Rc<()>)>`
-   = note: required because it appears within the type `hashbrown::map::HashMap<Rc<()>, Rc<()>, RandomState>`
-   = note: required because it appears within the type `HashMap<Rc<()>, Rc<()>>`
+note: required because it appears within the type `hashbrown::map::HashMap<Rc<()>, Rc<()>, RandomState>`
+  --> /cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.12.0/src/map.rs:188:12
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/test/ui/issues/issue-21763.stderr:11]: the old `def_ident_span` was using `get_if_local`, but the new logic works with defs from other crates, too

[the remark this answers is not in the document]
Comment (cjgillot) [reply src/test/ui/issues/issue-21763.stderr:11]: I mean: shouldn't this path be redacted somehow?

[the remark this answers is not in the document]
Comment (oli-obk) [reply src/test/ui/issues/issue-21763.stderr:11]: ah, well... we only do that for things in the /build directory, and this is actually fetched from crates.io.

so yea, we should teach compiletest about arbitrary directories that should be replaced in the stderr output

File: compiler/rustc_resolve/src/def_collector.rs [hunk 2]
```diff
@@ -283,13 +284,17 @@ impl<'a, 'b> visit::Visitor<'a> for DefCollector<'a, 'b> {
...
-                        DefPathData::ImplTrait,
-                        self.expansion.to_expn_id(),
-                        ty.span,
-                    ),
+                    ImplTraitContext::Universal(item_def) => {
+                        let name = rustc_span::Symbol::intern(&pprust::ty_to_string(ty));
+                        self.resolver.create_def(
+                            item_def,
+                            node_id,
+                            //DefPathData::ImplTrait,
```
Comment (oli-obk) [inline compiler/rustc_resolve/src/def_collector.rs:292]: Commented out code

File: src/test/ui/associated-type-bounds/assoc-type-eq-with-dyn-atb-fail.stderr [hunk 3]
```diff
@@ -4,7 +4,7 @@ error[E0277]: the trait bound `String: Copy` is not satisfied
 LL |         Box::new(AssocNoCopy)
    |         ^^^^^^^^^^^^^^^^^^^^^ the trait `Copy` is not implemented for `String`
    |
-   = note: required for the cast to the object type `dyn Bar<Assoc = <AssocNoCopy as Thing>::Out::{opaque#0}>`
+   = note: required for the cast to the object type `dyn Bar<Assoc = <AssocNoCopy as Thing>::Out::impl Copy>`
```
Comment (oli-obk) [inline src/test/ui/associated-type-bounds/assoc-type-eq-with-dyn-atb-fail.stderr:7]: This is somewhat suboptimal as you can't figure out which impl trait it is if there are multiple with the same bounds. Maybe leave out this change from this PR by adjusting path printing for opaque types?

**Review by oli-obk — CHANGES_REQUESTED** [review]
I'll look into compiletest changes next week. I wanted to add a way to add defaults for the comment-configs of ui tests anyway, and we could use that to insert a filter for all crates.io crate paths

File: src/test/ui/issues/issue-21763.stderr (hunk 1 above)
[the remark this answers is not in the document]
Comment (cjgillot) [reply src/test/ui/issues/issue-21763.stderr]: I removed the generalized use of `def_ident_span` from this PR.

**Pull request merged**